haskell/src/SwiftNav/SBP.hs

```diff
@@ -0,0 +1,124 @@
+module SwiftNav.SBP where
+
+import Data.Binary
+import Data.Binary.Get
+import Data.Binary.Put
+import Data.ByteString
+import Data.Word
+import SwiftNav.SBP.Acquisition
+import SwiftNav.SBP.Bootload
+import SwiftNav.SBP.ExtEvents
+import SwiftNav.SBP.FileIo
+import SwiftNav.SBP.Flash
+import SwiftNav.SBP.Logging
+import SwiftNav.SBP.Navigation
+import SwiftNav.SBP.Observation
+import SwiftNav.SBP.Piksi
+import SwiftNav.SBP.Settings
+import SwiftNav.SBP.System
+import SwiftNav.SBP.Tracking
+
+msgPreamble :: Word8
+msgPreamble = 0x55
+
+data Msg = Msg
+  { msgSBPType    :: Word16
+  , msgSBPSender  :: Word16
+  , msgSBPLen     :: Word8
+  , msgSBPPayload :: ByteString
+  , msgSBPCrc     :: Word16
+  } deriving ( Show, Read, Eq )
+
+instance Binary Msg where
+  get = do
+    msgSBPType <- getWord16le
+    msgSBPSender <- getWord16le
+    msgSBPLen <- getWord8
+    msgSBPPayload <- getByteString $ fromIntegral msgSBPLen
+    msgSBPCrc <- getWord16le
+    return Msg {..}
+
+  put Msg {..} = do
+    putWord16le msgSBPType
+    putWord16le msgSBPSender
+    putWord8 msgSBPLen
+    putByteString msgSBPPayload
+    putWord16le msgSBPCrc
+
+getMsg :: Get Msg
+getMsg = do
+  preamble <- getWord8
+  if preamble /= msgPreamble then getMsg else get
+
+putMsg :: Msg -> Put
+putMsg msg = do
+  putWord8 msgPreamble
+  put msg
+
+data SBPMsg =
```

Comment:
I may have said this before, this style turns into an "expression problem"-type concern. Ideally, you'd want to be able to define and extend messages through a uniform interface of some kind. I suppose we have a limited and well-defined set of messages, so it's probably not that bad.